// ionic-angular/components/nav/nav-push.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from 'ionic-angular/components/nav/nav-push';
import * as import1 from '@angular/core/src/change_detection/change_detection';
import * as import2 from '@angular/core/src/linker/view_utils';
import * as import3 from '@angular/core/src/linker/view';
export class Wrapper_NavPush {
  context:import0.NavPush;
  changed:boolean;
  /*private*/ _navPush:any;
  /*private*/ _navParams:any;
  constructor(p0:any) {
    this.changed = false;
    this._navPush = import1.UNINITIALIZED;
    this._navParams = import1.UNINITIALIZED;
    this.context = new import0.NavPush(p0);
  }
  check_navPush(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._navPush,currValue))) {
      this.changed = true;
      this.context.navPush = currValue;
      this._navPush = currValue;
    }
  }
  check_navParams(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._navParams,currValue))) {
      this.changed = true;
      this.context.navParams = currValue;
      this._navParams = currValue;
    }
  }
  detectChangesInternal(view:import3.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
export class Wrapper_NavPushAnchor {
  context:import0.NavPushAnchor;
  changed:boolean;
  constructor(p0:any,p1:any) {
    this.changed = false;
    this.context = new import0.NavPushAnchor(p0,p1);
  }
  detectChangesInternal(view:import3.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
