// @ionic/app-scripts/dist/closure.js
"use strict";
var config_1 = require('./util/config');
var logger_1 = require('./util/logger');
var worker_client_1 = require('./worker-client');
function closure(context, configFile) {
    context = config_1.generateContext(context);
    configFile = config_1.getUserConfigFile(context, taskInfo, configFile);
    var logger = new logger_1.Logger('closure');
    return worker_client_1.runWorker('closure', 'closureWorker', context, configFile)
        .then(function () {
        logger.finish();
    })
        .catch(function (err) {
        throw logger.fail(err);
    });
}
exports.closure = closure;
function closureWorker(context, configFile) {
    return new Promise(function (resolve, reject) {
        logger_1.Logger.warn('Closer Compiler unsupported at this time.');
        resolve();
    });
}
exports.closureWorker = closureWorker;
function isClosureSupported(context) {
    // TODO: check for Java and compiler.jar executable
    return false;
}
exports.isClosureSupported = isClosureSupported;
var taskInfo = {
    fullArgConfig: '--closure',
    shortArgConfig: '-l',
    envConfig: 'ionic_closure',
    defaultConfigFile: 'closure.config'
};
