// @ionic/app-scripts/dist/copy.js
"use strict";
var logger_1 = require('./util/logger');
var events_1 = require('./util/events');
var config_1 = require('./util/config');
var fs = require('fs-extra');
var path_1 = require('path');
function copy(context, configFile) {
    context = config_1.generateContext(context);
    configFile = config_1.getUserConfigFile(context, taskInfo, configFile);
    var logger = new logger_1.Logger('copy');
    return copyWorker(context, configFile)
        .then(function () {
        logger.finish();
    })
        .catch(function (err) {
        throw logger.fail(err);
    });
}
exports.copy = copy;
function copyUpdate(event, filePath, context) {
    var configFile = config_1.getUserConfigFile(context, taskInfo, null);
    logger_1.Logger.debug("copyUpdate, event: " + event + ", path: " + filePath);
    if (event === 'change' || event === 'add' || event === 'addDir') {
        // figure out which copy option(s) this one file/directory belongs to
        var copyConfig = config_1.fillConfigDefaults(configFile, taskInfo.defaultConfigFile);
        var fileCopyOptions = findFileCopyOptions(context, copyConfig, filePath);
        if (fileCopyOptions.length) {
            var promises = fileCopyOptions.map(function (copyOptions) {
                return copySrcToDest(context, copyOptions.src, copyOptions.dest, copyOptions.filter, true);
            });
            return Promise.all(promises).then(function (copySrcToDestResults) {
                printCopyErrorMessages('copy', copySrcToDestResults);
                var destFiles = copySrcToDestResults.map(function (copySrcToDestResult) { return copySrcToDestResult.dest; });
                events_1.emit(events_1.EventType.FileChange, destFiles);
            });
        }
    }
    else if (event === 'unlink' || event === 'unlinkDir') {
        return new Promise(function (resolve, reject) {
            var destFile = path_1.join(context.rootDir, filePath);
            fs.remove(destFile, function (err) {
                if (err) {
                    reject(new logger_1.BuildError(err));
                }
                else {
                    if (event === 'unlink') {
                        events_1.emit(events_1.EventType.FileDelete, destFile);
                    }
                    else if (event === 'unlinkDir') {
                        events_1.emit(events_1.EventType.DirectoryDelete, destFile);
                    }
                    resolve();
                }
            });
        });
    }
    return copyWorker(context, configFile);
}
exports.copyUpdate = copyUpdate;
function copyWorker(context, configFile) {
    var copyConfig = config_1.fillConfigDefaults(configFile, taskInfo.defaultConfigFile);
    var promises = copyConfig.include.map(function (copyOptions) {
        return copySrcToDest(context, copyOptions.src, copyOptions.dest, copyOptions.filter, true);
    });
    return Promise.all(promises).then(function (copySrcToDestResults) {
        printCopyErrorMessages('copy', copySrcToDestResults);
    });
}
exports.copyWorker = copyWorker;
function printCopyErrorMessages(prefix, copyResults) {
    copyResults.forEach(function (copyResult) {
        if (!copyResult.success) {
            logger_1.Logger.warn(prefix + ": " + copyResult.errorMessage);
        }
    });
}
function findFileCopyOptions(context, copyConfig, filePath) {
    var copyOptions = [];
    if (!copyConfig || !copyConfig.include || !context.rootDir) {
        return copyOptions;
    }
    filePath = path_1.join(context.rootDir, filePath);
    var filePathSegments = filePath.split(path_1.sep);
    var srcPath;
    var srcLookupPath;
    var destCopyOption;
    var destPath;
    while (filePathSegments.length > 1) {
        srcPath = filePathSegments.join(path_1.sep);
        for (var i = 0; i < copyConfig.include.length; i++) {
            srcLookupPath = path_1.resolve(config_1.replacePathVars(context, copyConfig.include[i].src));
            if (srcPath === srcLookupPath) {
                destCopyOption = path_1.resolve(config_1.replacePathVars(context, copyConfig.include[i].dest));
                destPath = filePath.replace(srcLookupPath, destCopyOption);
                copyOptions.push({
                    src: filePath,
                    dest: destPath,
                    filter: copyConfig.include[i].filter
                });
            }
        }
        if (srcPath.length < context.rootDir.length) {
            break;
        }
        filePathSegments.pop();
    }
    return copyOptions;
}
exports.findFileCopyOptions = findFileCopyOptions;
function copySrcToDest(context, src, dest, filter, clobber) {
    return new Promise(function (resolve, reject) {
        src = path_1.resolve(config_1.replacePathVars(context, src));
        dest = path_1.resolve(config_1.replacePathVars(context, dest));
        var opts = {
            filter: filter,
            clobber: clobber
        };
        fs.copy(src, dest, opts, function (err) {
            if (err) {
                if (err.message.indexOf('ENOENT') > -1) {
                    resolve({ success: false, src: src, dest: dest, errorMessage: "Error copying \"" + src + "\" to \"" + dest + "\": File not found" });
                }
                else {
                    resolve({ success: false, src: src, dest: dest, errorMessage: "Error copying \"" + src + "\" to \"" + dest + "\"" });
                }
                return;
            }
            resolve({ success: true, src: src, dest: dest });
        });
    });
}
var taskInfo = {
    fullArgConfig: '--copy',
    shortArgConfig: '-y',
    envConfig: 'ionic_copy',
    defaultConfigFile: 'copy.config'
};
