// @ionic/app-scripts/dist/ngc.js
"use strict";
var path_1 = require('path');
var fs_extra_1 = require('fs-extra');
var helpers_1 = require('./util/helpers');
var config_1 = require('./util/config');
var transpile_1 = require('./transpile');
var logger_1 = require('./util/logger');
var ts = require('typescript');
function ngc(context, configFile) {
    context = config_1.generateContext(context);
    configFile = config_1.getUserConfigFile(context, taskInfo, configFile);
    var logger = new logger_1.Logger('ngc');
    return ngcWorker(context, configFile)
        .then(function () {
        logger.finish();
    })
        .catch(function (err) {
        throw logger.fail(err);
    });
}
exports.ngc = ngc;
function ngcWorker(context, configFile) {
    // first make a copy of src TS files
    // and copy them into the tmp directory
    return copySrcTsToTmpDir(context).then(function () {
        return runNgc(context, configFile);
    });
}
exports.ngcWorker = ngcWorker;
function runNgc(context, configFile) {
    return new Promise(function (resolve, reject) {
        var ngcConfig = config_1.fillConfigDefaults(configFile, taskInfo.defaultConfigFile);
        // make a copy of the users src tsconfig file
        // and save the modified copy into the tmp directory
        createTmpTsConfig(context, ngcConfig);
        var ngcCmd = config_1.getNodeBinExecutable(context, 'ngc');
        if (!ngcCmd) {
            reject(new logger_1.BuildError("Unable to find Angular Compiler \"ngc\" command: " + ngcCmd + ". Please ensure @angular/compiler-cli has been installed with NPM."));
            return;
        }
        // let's kick off the actual ngc command on our copied TS files
        // use the user's ngc in their node_modules to ensure ngc
        // versioned and working along with the user's ng2 version
        var spawn = require('cross-spawn');
        var ngcCmdArgs = [
            '--project', getTmpTsConfigPath(context)
        ];
        logger_1.Logger.debug("run: " + ngcCmd + " " + ngcCmdArgs.join(' '));
        // would love to not use spawn here but import and run ngc directly
        var cp = spawn(ngcCmd, ngcCmdArgs);
        var errorMsgs = [];
        cp.stdout.on('data', function (data) {
            logger_1.Logger.info(data);
        });
        cp.stderr.on('data', function (data) {
            if (data) {
                data.toString().split('\n').forEach(function (line) {
                    if (!line.trim().length) {
                        // if it's got no data then don't bother
                        return;
                    }
                    if (line.substr(0, 4) === '    ' || line === 'Compilation failed') {
                        // if it's indented then it's some callstack message we don't care about
                        return;
                    }
                    // split by the : character, then rebuild the line until it's too long
                    // and make a new line
                    var lineSections = line.split(': ');
                    var msgSections = [];
                    for (var i = 0; i < lineSections.length; i++) {
                        msgSections.push(lineSections[i]);
                        if (msgSections.join(': ').length > 40) {
                            errorMsgs.push(msgSections.join(': '));
                            msgSections = [];
                        }
                    }
                    if (msgSections.length) {
                        errorMsgs.push(msgSections.join(': '));
                    }
                });
            }
        });
        cp.on('close', function (code) {
            if (errorMsgs.length) {
                errorMsgs.forEach(function (errorMsg) {
                    logger_1.Logger.error(errorMsg);
                });
                reject(new logger_1.BuildError());
            }
            else {
                resolve();
            }
        });
    });
}
function createTmpTsConfig(context, ngcConfig) {
    // create the tsconfig from the original src
    var tsConfigPath = transpile_1.getTsConfigPath(context);
    var tsConfigFile = ts.readConfigFile(tsConfigPath, function (path) { return fs_extra_1.readFileSync(path, 'utf8'); });
    if (!tsConfigFile || !tsConfigFile.config) {
        throw new logger_1.BuildError("invalid tsconfig: " + tsConfigPath);
    }
    if (!tsConfigFile.config.compilerOptions) {
        throw new logger_1.BuildError("invalid tsconfig compilerOptions: " + tsConfigPath);
    }
    // delete outDir if it's set since we only want
    // to compile to the same directory we're in
    delete tsConfigFile.config.compilerOptions.outDir;
    var mergedConfig = helpers_1.objectAssign({}, tsConfigFile.config, ngcConfig);
    // save the modified copy into the tmp directory
    fs_extra_1.outputJsonSync(getTmpTsConfigPath(context), mergedConfig);
}
function copySrcTsToTmpDir(context) {
    return new Promise(function (resolve, reject) {
        // ensure the tmp directory is ready to go
        try {
            fs_extra_1.emptyDirSync(context.tmpDir);
        }
        catch (e) {
            reject(new logger_1.BuildError("tmpDir error: " + e));
            return;
        }
        var copyOpts = {
            filter: filterCopyFiles
        };
        logger_1.Logger.debug("copySrcTsToTmpDir, srcDir: " + context.srcDir + " to tmpDir: " + context.tmpDir);
        fs_extra_1.copy(context.srcDir, context.tmpDir, copyOpts, function (err) {
            if (err) {
                reject(new logger_1.BuildError(err));
            }
            else {
                resolve();
            }
        });
    });
}
function filterCopyFiles(filePath, hoop) {
    var shouldInclude = false;
    try {
        var stats = fs_extra_1.statSync(filePath);
        if (stats.isDirectory()) {
            shouldInclude = (EXCLUDE_DIRS.indexOf(path_1.basename(filePath)) < 0);
        }
        else {
            shouldInclude = (helpers_1.endsWith(filePath, '.ts') || helpers_1.endsWith(filePath, '.html'));
        }
    }
    catch (e) { }
    return shouldInclude;
}
function getTmpTsConfigPath(context) {
    return path_1.join(context.tmpDir, 'tsconfig.json');
}
exports.getTmpTsConfigPath = getTmpTsConfigPath;
var EXCLUDE_DIRS = ['assets', 'theme'];
var taskInfo = {
    fullArgConfig: '--ngc',
    shortArgConfig: '-n',
    envConfig: 'ionic_ngc',
    defaultConfigFile: 'ngc.config'
};
