// ionic-angular/components/grid/grid.ngfactory.js
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
import * as import0 from 'ionic-angular/components/grid/grid';
export var Wrapper_Grid = (function () {
    function Wrapper_Grid() {
        this.changed = false;
        this.context = new import0.Grid();
    }
    Wrapper_Grid.prototype.detectChangesInternal = function (view, el, throwOnChange) {
        var changed = this.changed;
        this.changed = false;
        return changed;
    };
    return Wrapper_Grid;
}());
export var Wrapper_Row = (function () {
    function Wrapper_Row() {
        this.changed = false;
        this.context = new import0.Row();
    }
    Wrapper_Row.prototype.detectChangesInternal = function (view, el, throwOnChange) {
        var changed = this.changed;
        this.changed = false;
        return changed;
    };
    return Wrapper_Row;
}());
export var Wrapper_Col = (function () {
    function Wrapper_Col() {
        this.changed = false;
        this.context = new import0.Col();
    }
    Wrapper_Col.prototype.detectChangesInternal = function (view, el, throwOnChange) {
        var changed = this.changed;
        this.changed = false;
        return changed;
    };
    return Wrapper_Col;
}());
