// ionic-angular/components/grid/grid.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from 'ionic-angular/components/grid/grid';
import * as import1 from '@angular/core/src/linker/view';
export class Wrapper_Grid {
  context:import0.Grid;
  changed:boolean;
  constructor() {
    this.changed = false;
    this.context = new import0.Grid();
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
export class Wrapper_Row {
  context:import0.Row;
  changed:boolean;
  constructor() {
    this.changed = false;
    this.context = new import0.Row();
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
export class Wrapper_Col {
  context:import0.Col;
  changed:boolean;
  constructor() {
    this.changed = false;
    this.context = new import0.Col();
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
