// @ionic/app-scripts/dist/util/logger-typescript.js
"use strict";
var logger_1 = require('./logger');
var helpers_1 = require('./helpers');
var ts = require('typescript');
/**
 * Ok, so formatting overkill, we know. But whatever, it makes for great
 * error reporting within a terminal. So, yeah, let's code it up, shall we?
 */
function runDiagnostics(context, tsDiagnostics) {
    var diagnostics = tsDiagnostics.map(function (tsDiagnostic) {
        return loadDiagnostic(context, tsDiagnostic);
    });
    if (diagnostics.length > 0) {
        diagnostics.forEach(function (d) {
            logger_1.Logger.printDiagnostic(helpers_1.objectAssign({}, d));
        });
        return true;
    }
    return false;
}
exports.runDiagnostics = runDiagnostics;
function loadDiagnostic(context, tsDiagnostic) {
    var d = {
        level: 'error',
        syntax: 'js',
        type: 'typescript',
        header: 'typescript error',
        code: tsDiagnostic.code.toString(),
        messageText: ts.flattenDiagnosticMessageText(tsDiagnostic.messageText, '\n'),
        fileName: null,
        lines: []
    };
    if (tsDiagnostic.file) {
        d.fileName = logger_1.Logger.formatFileName(context.rootDir, tsDiagnostic.file.fileName);
        var srcLines = tsDiagnostic.file.getText().replace(/\\r/g, '\n').split('\n');
        var posData = tsDiagnostic.file.getLineAndCharacterOfPosition(tsDiagnostic.start);
        var errorLine = {
            lineIndex: posData.line,
            lineNumber: posData.line + 1,
            text: srcLines[posData.line],
            errorCharStart: posData.character,
            errorLength: Math.max(tsDiagnostic.length, 1)
        };
        d.lines.push(errorLine);
        if (errorLine.errorLength === 0 && errorLine.errorCharStart > 0) {
            errorLine.errorLength = 1;
            errorLine.errorCharStart--;
        }
        d.header = logger_1.Logger.formatHeader('typescript', tsDiagnostic.file.fileName, context.rootDir, errorLine.lineNumber);
        if (errorLine.lineIndex > 0 && logger_1.Logger.meaningfulLine(srcLines[errorLine.lineIndex - 1])) {
            var previousLine = {
                lineIndex: errorLine.lineIndex - 1,
                lineNumber: errorLine.lineNumber - 1,
                text: srcLines[errorLine.lineIndex - 1],
                errorCharStart: -1,
                errorLength: -1
            };
            d.lines.unshift(previousLine);
        }
        if (errorLine.lineIndex + 1 < srcLines.length && logger_1.Logger.meaningfulLine(srcLines[errorLine.lineIndex + 1])) {
            var nextLine = {
                lineIndex: errorLine.lineIndex + 1,
                lineNumber: errorLine.lineNumber + 1,
                text: srcLines[errorLine.lineIndex + 1],
                errorCharStart: -1,
                errorLength: -1
            };
            d.lines.push(nextLine);
        }
    }
    return d;
}
