// ionic-angular/components/list/list.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from 'ionic-angular/components/list/list';
import * as import1 from '@angular/core/src/change_detection/change_detection';
import * as import2 from '@angular/core/src/linker/view_utils';
import * as import3 from '@angular/core/src/linker/view';
export class Wrapper_List {
  context:import0.List;
  changed:boolean;
  /*private*/ _mode:any;
  /*private*/ _sliding:any;
  constructor(p0:any,p1:any,p2:any,p3:any) {
    this.changed = false;
    this._mode = import1.UNINITIALIZED;
    this._sliding = import1.UNINITIALIZED;
    this.context = new import0.List(p0,p1,p2,p3);
  }
  check_mode(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._mode,currValue))) {
      this.changed = true;
      this.context.mode = currValue;
      this._mode = currValue;
    }
  }
  check_sliding(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import2.checkBinding(throwOnChange,this._sliding,currValue))) {
      this.changed = true;
      this.context.sliding = currValue;
      this._sliding = currValue;
    }
  }
  detectChangesInternal(view:import3.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
