// ionic-angular/components/action-sheet/action-sheet-component.ngfactory.js
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import * as import0 from 'ionic-angular/components/action-sheet/action-sheet-component';
import * as import1 from '@angular/core/src/linker/view';
import * as import3 from '@angular/core/src/linker/element';
import * as import4 from '@angular/core/src/linker/view_utils';
import * as import6 from '@angular/core/src/linker/view_type';
import * as import7 from '@angular/core/src/change_detection/change_detection';
import * as import8 from 'ionic-angular/navigation/view-controller';
import * as import9 from 'ionic-angular/config/config';
import * as import10 from '@angular/core/src/linker/element_ref';
import * as import11 from 'ionic-angular/util/form';
import * as import12 from 'ionic-angular/navigation/nav-params';
import * as import13 from '@angular/core/src/metadata/view';
import * as import14 from '@angular/core/src/linker/component_factory';
import * as import15 from '../backdrop/backdrop.ngfactory';
import * as import16 from '../../../@angular/common/src/directives/ng_if.ngfactory';
import * as import17 from '../../../@angular/common/src/directives/ng_for.ngfactory';
import * as import18 from 'ionic-angular/gestures/gesture-controller';
import * as import19 from '@angular/core/src/linker/template_ref';
import * as import20 from '@angular/core/src/change_detection/differs/iterable_differs';
import * as import21 from 'ionic-angular/components/backdrop/backdrop';
import * as import22 from '@angular/common/src/directives/ng_if';
import * as import23 from '@angular/common/src/directives/ng_for';
import * as import24 from '../../../@angular/common/src/directives/ng_class.ngfactory';
import * as import25 from '../button/button.ngfactory';
import * as import26 from '@angular/core/src/change_detection/differs/keyvalue_differs';
import * as import27 from '@angular/common/src/directives/ng_class';
import * as import28 from 'ionic-angular/components/button/button';
import * as import29 from '../icon/icon.ngfactory';
import * as import30 from 'ionic-angular/components/icon/icon';
export var Wrapper_ActionSheetCmp = (function () {
    function Wrapper_ActionSheetCmp(p0, p1, p2, p3, p4, p5) {
        this.changed = false;
        this.context = new import0.ActionSheetCmp(p0, p1, p2, p3, p4, p5);
    }
    Wrapper_ActionSheetCmp.prototype.detectChangesInternal = function (view, el, throwOnChange) {
        var changed = this.changed;
        this.changed = false;
        return changed;
    };
    return Wrapper_ActionSheetCmp;
}());
var renderType_ActionSheetCmp_Host = null;
var _View_ActionSheetCmp_Host0 = (function (_super) {
    __extends(_View_ActionSheetCmp_Host0, _super);
    function _View_ActionSheetCmp_Host0(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp_Host0, renderType_ActionSheetCmp_Host, import6.ViewType.HOST, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp_Host0.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.selectOrCreateHostElement('ion-action-sheet', rootSelector, null);
        this.renderer.setElementAttribute(this._el_0, 'role', 'dialog');
        this._appEl_0 = new import3.AppElement(0, null, this, this._el_0);
        var compView_0 = viewFactory_ActionSheetCmp0(this.viewUtils, this.injector(0), this._appEl_0);
        this._ActionSheetCmp_0_4 = new Wrapper_ActionSheetCmp(this.parentInjector.get(import8.ViewController), this.parentInjector.get(import9.Config), new import10.ElementRef(this._el_0), this.parentInjector.get(import11.Form), this.parentInjector.get(import12.NavParams), this.renderer);
        this._appEl_0.initComponent(this._ActionSheetCmp_0_4.context, [], compView_0);
        compView_0.create(this._ActionSheetCmp_0_4.context, this.projectableNodes, null);
        var disposable_0 = this.renderer.listenGlobal('body', 'keyup', this.eventHandler(this._handle_keyup_0_0.bind(this)));
        this._expr_1 = import7.UNINITIALIZED;
        this._expr_2 = import7.UNINITIALIZED;
        this.init([].concat([this._el_0]), [this._el_0], [disposable_0], []);
        return this._appEl_0;
    };
    _View_ActionSheetCmp_Host0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import0.ActionSheetCmp) && (0 === requestNodeIndex))) {
            return this._ActionSheetCmp_0_4.context;
        }
        return notFoundResult;
    };
    _View_ActionSheetCmp_Host0.prototype.detectChangesInternal = function (throwOnChange) {
        this._ActionSheetCmp_0_4.detectChangesInternal(this, this._el_0, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        var currVal_1 = this._ActionSheetCmp_0_4.context.hdrId;
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setElementAttribute(this._el_0, 'aria-labelledby', ((currVal_1 == null) ? null : currVal_1.toString()));
            this._expr_1 = currVal_1;
        }
        var currVal_2 = this._ActionSheetCmp_0_4.context.descId;
        if (import4.checkBinding(throwOnChange, this._expr_2, currVal_2)) {
            this.renderer.setElementAttribute(this._el_0, 'aria-describedby', ((currVal_2 == null) ? null : currVal_2.toString()));
            this._expr_2 = currVal_2;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ActionSheetCmp_Host0.prototype._handle_keyup_0_0 = function ($event) {
        this._appEl_0.componentView.markPathToRootAsCheckOnce();
        var pd_0 = (this._ActionSheetCmp_0_4.context.keyUp($event) !== false);
        return (true && pd_0);
    };
    return _View_ActionSheetCmp_Host0;
}(import1.AppView));
function viewFactory_ActionSheetCmp_Host0(viewUtils, parentInjector, declarationEl) {
    if ((renderType_ActionSheetCmp_Host === null)) {
        (renderType_ActionSheetCmp_Host = viewUtils.createRenderComponentType('', 0, import13.ViewEncapsulation.None, [], {}));
    }
    return new _View_ActionSheetCmp_Host0(viewUtils, parentInjector, declarationEl);
}
export var ActionSheetCmpNgFactory = new import14.ComponentFactory('ion-action-sheet', viewFactory_ActionSheetCmp_Host0, import0.ActionSheetCmp);
var styles_ActionSheetCmp = [];
var renderType_ActionSheetCmp = null;
var _View_ActionSheetCmp0 = (function (_super) {
    __extends(_View_ActionSheetCmp0, _super);
    function _View_ActionSheetCmp0(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp0, renderType_ActionSheetCmp, import6.ViewType.COMPONENT, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp0.prototype.createInternal = function (rootSelector) {
        var parentRenderNode = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
        this._el_0 = this.renderer.createElement(parentRenderNode, 'ion-backdrop', null);
        this.renderer.setElementAttribute(this._el_0, 'disable-activated', '');
        this.renderer.setElementAttribute(this._el_0, 'role', 'presentation');
        this.renderer.setElementAttribute(this._el_0, 'tappable', '');
        this._Backdrop_0_3 = new import15.Wrapper_Backdrop(this.parentInjector.get(import18.GestureController), new import10.ElementRef(this._el_0), this.renderer);
        this._el_1 = this.renderer.createElement(parentRenderNode, 'div', null);
        this.renderer.setElementAttribute(this._el_1, 'class', 'action-sheet-wrapper');
        this._el_2 = this.renderer.createElement(this._el_1, 'div', null);
        this.renderer.setElementAttribute(this._el_2, 'class', 'action-sheet-container');
        this._el_3 = this.renderer.createElement(this._el_2, 'div', null);
        this.renderer.setElementAttribute(this._el_3, 'class', 'action-sheet-group');
        this._anchor_4 = this.renderer.createTemplateAnchor(this._el_3, null);
        this._appEl_4 = new import3.AppElement(4, 3, this, this._anchor_4);
        this._TemplateRef_4_5 = new import19.TemplateRef_(this._appEl_4, viewFactory_ActionSheetCmp1);
        this._NgIf_4_6 = new import16.Wrapper_NgIf(this._appEl_4.vcRef, this._TemplateRef_4_5);
        this._anchor_5 = this.renderer.createTemplateAnchor(this._el_3, null);
        this._appEl_5 = new import3.AppElement(5, 3, this, this._anchor_5);
        this._TemplateRef_5_5 = new import19.TemplateRef_(this._appEl_5, viewFactory_ActionSheetCmp2);
        this._NgIf_5_6 = new import16.Wrapper_NgIf(this._appEl_5.vcRef, this._TemplateRef_5_5);
        this._anchor_6 = this.renderer.createTemplateAnchor(this._el_3, null);
        this._appEl_6 = new import3.AppElement(6, 3, this, this._anchor_6);
        this._TemplateRef_6_5 = new import19.TemplateRef_(this._appEl_6, viewFactory_ActionSheetCmp3);
        this._NgFor_6_6 = new import17.Wrapper_NgFor(this._appEl_6.vcRef, this._TemplateRef_6_5, this.parentInjector.get(import20.IterableDiffers), this.ref);
        this._anchor_7 = this.renderer.createTemplateAnchor(this._el_2, null);
        this._appEl_7 = new import3.AppElement(7, 2, this, this._anchor_7);
        this._TemplateRef_7_5 = new import19.TemplateRef_(this._appEl_7, viewFactory_ActionSheetCmp5);
        this._NgIf_7_6 = new import16.Wrapper_NgIf(this._appEl_7.vcRef, this._TemplateRef_7_5);
        var disposable_0 = this.renderer.listen(this._el_0, 'click', this.eventHandler(this._handle_click_0_0.bind(this)));
        this.init([], [
            this._el_0,
            this._el_1,
            this._el_2,
            this._el_3,
            this._anchor_4,
            this._anchor_5,
            this._anchor_6,
            this._anchor_7
        ], [disposable_0], []);
        return null;
    };
    _View_ActionSheetCmp0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import21.Backdrop) && (0 === requestNodeIndex))) {
            return this._Backdrop_0_3.context;
        }
        if (((token === import19.TemplateRef) && (4 === requestNodeIndex))) {
            return this._TemplateRef_4_5;
        }
        if (((token === import22.NgIf) && (4 === requestNodeIndex))) {
            return this._NgIf_4_6.context;
        }
        if (((token === import19.TemplateRef) && (5 === requestNodeIndex))) {
            return this._TemplateRef_5_5;
        }
        if (((token === import22.NgIf) && (5 === requestNodeIndex))) {
            return this._NgIf_5_6.context;
        }
        if (((token === import19.TemplateRef) && (6 === requestNodeIndex))) {
            return this._TemplateRef_6_5;
        }
        if (((token === import23.NgFor) && (6 === requestNodeIndex))) {
            return this._NgFor_6_6.context;
        }
        if (((token === import19.TemplateRef) && (7 === requestNodeIndex))) {
            return this._TemplateRef_7_5;
        }
        if (((token === import22.NgIf) && (7 === requestNodeIndex))) {
            return this._NgIf_7_6.context;
        }
        return notFoundResult;
    };
    _View_ActionSheetCmp0.prototype.detectChangesInternal = function (throwOnChange) {
        this._Backdrop_0_3.detectChangesInternal(this, this._el_0, throwOnChange);
        var currVal_1 = this.context.d.title;
        this._NgIf_4_6.check_ngIf(currVal_1, throwOnChange, false);
        this._NgIf_4_6.detectChangesInternal(this, this._anchor_4, throwOnChange);
        var currVal_2 = this.context.d.subTitle;
        this._NgIf_5_6.check_ngIf(currVal_2, throwOnChange, false);
        this._NgIf_5_6.detectChangesInternal(this, this._anchor_5, throwOnChange);
        var currVal_3 = this.context.d.buttons;
        this._NgFor_6_6.check_ngForOf(currVal_3, throwOnChange, false);
        this._NgFor_6_6.detectChangesInternal(this, this._anchor_6, throwOnChange);
        var currVal_4 = this.context.d.cancelButton;
        this._NgIf_7_6.check_ngIf(currVal_4, throwOnChange, false);
        this._NgIf_7_6.detectChangesInternal(this, this._anchor_7, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ActionSheetCmp0.prototype.destroyInternal = function () {
        this._Backdrop_0_3.context.ngOnDestroy();
    };
    _View_ActionSheetCmp0.prototype._handle_click_0_0 = function ($event) {
        this.markPathToRootAsCheckOnce();
        var pd_0 = (this.context.bdClick() !== false);
        return (true && pd_0);
    };
    return _View_ActionSheetCmp0;
}(import1.AppView));
export function viewFactory_ActionSheetCmp0(viewUtils, parentInjector, declarationEl) {
    if ((renderType_ActionSheetCmp === null)) {
        (renderType_ActionSheetCmp = viewUtils.createRenderComponentType('', 0, import13.ViewEncapsulation.None, styles_ActionSheetCmp, {}));
    }
    return new _View_ActionSheetCmp0(viewUtils, parentInjector, declarationEl);
}
var _View_ActionSheetCmp1 = (function (_super) {
    __extends(_View_ActionSheetCmp1, _super);
    function _View_ActionSheetCmp1(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp1, renderType_ActionSheetCmp, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp1.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.renderer.createElement(null, 'div', null);
        this.renderer.setElementAttribute(this._el_0, 'class', 'action-sheet-title');
        this._text_1 = this.renderer.createText(this._el_0, '', null);
        this._expr_0 = import7.UNINITIALIZED;
        this._expr_1 = import7.UNINITIALIZED;
        this.init([].concat([this._el_0]), [
            this._el_0,
            this._text_1
        ], [], []);
        return null;
    };
    _View_ActionSheetCmp1.prototype.detectChangesInternal = function (throwOnChange) {
        this.detectContentChildrenChanges(throwOnChange);
        var currVal_0 = import4.interpolate(1, '', this.parent.context.hdrId, '');
        if (import4.checkBinding(throwOnChange, this._expr_0, currVal_0)) {
            this.renderer.setElementProperty(this._el_0, 'id', currVal_0);
            this._expr_0 = currVal_0;
        }
        var currVal_1 = import4.interpolate(1, '', this.parent.context.d.title, '');
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setText(this._text_1, currVal_1);
            this._expr_1 = currVal_1;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    return _View_ActionSheetCmp1;
}(import1.AppView));
function viewFactory_ActionSheetCmp1(viewUtils, parentInjector, declarationEl) {
    return new _View_ActionSheetCmp1(viewUtils, parentInjector, declarationEl);
}
var _View_ActionSheetCmp2 = (function (_super) {
    __extends(_View_ActionSheetCmp2, _super);
    function _View_ActionSheetCmp2(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp2, renderType_ActionSheetCmp, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp2.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.renderer.createElement(null, 'div', null);
        this.renderer.setElementAttribute(this._el_0, 'class', 'action-sheet-sub-title');
        this._text_1 = this.renderer.createText(this._el_0, '', null);
        this._expr_0 = import7.UNINITIALIZED;
        this._expr_1 = import7.UNINITIALIZED;
        this.init([].concat([this._el_0]), [
            this._el_0,
            this._text_1
        ], [], []);
        return null;
    };
    _View_ActionSheetCmp2.prototype.detectChangesInternal = function (throwOnChange) {
        this.detectContentChildrenChanges(throwOnChange);
        var currVal_0 = import4.interpolate(1, '', this.parent.context.descId, '');
        if (import4.checkBinding(throwOnChange, this._expr_0, currVal_0)) {
            this.renderer.setElementProperty(this._el_0, 'id', currVal_0);
            this._expr_0 = currVal_0;
        }
        var currVal_1 = import4.interpolate(1, '', this.parent.context.d.subTitle, '');
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setText(this._text_1, currVal_1);
            this._expr_1 = currVal_1;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    return _View_ActionSheetCmp2;
}(import1.AppView));
function viewFactory_ActionSheetCmp2(viewUtils, parentInjector, declarationEl) {
    return new _View_ActionSheetCmp2(viewUtils, parentInjector, declarationEl);
}
var _View_ActionSheetCmp3 = (function (_super) {
    __extends(_View_ActionSheetCmp3, _super);
    function _View_ActionSheetCmp3(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp3, renderType_ActionSheetCmp, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp3.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.renderer.createElement(null, 'button', null);
        this.renderer.setElementAttribute(this._el_0, 'class', 'disable-hover');
        this.renderer.setElementAttribute(this._el_0, 'ion-button', 'action-sheet-button');
        this._appEl_0 = new import3.AppElement(0, null, this, this._el_0);
        var compView_0 = import25.viewFactory_Button0(this.viewUtils, this.injector(0), this._appEl_0);
        this._NgClass_0_4 = new import24.Wrapper_NgClass(this.parent.parentInjector.get(import20.IterableDiffers), this.parent.parentInjector.get(import26.KeyValueDiffers), new import10.ElementRef(this._el_0), this.renderer);
        this._Button_0_5 = new import25.Wrapper_Button(null, 'action-sheet-button', this.parent.parentInjector.get(import9.Config), new import10.ElementRef(this._el_0), this.renderer);
        this._appEl_0.initComponent(this._Button_0_5.context, [], compView_0);
        this._anchor_1 = this.renderer.createTemplateAnchor(null, null);
        this._appEl_1 = new import3.AppElement(1, 0, this, this._anchor_1);
        this._TemplateRef_1_5 = new import19.TemplateRef_(this._appEl_1, viewFactory_ActionSheetCmp4);
        this._NgIf_1_6 = new import16.Wrapper_NgIf(this._appEl_1.vcRef, this._TemplateRef_1_5);
        this._text_2 = this.renderer.createText(null, '', null);
        compView_0.create(this._Button_0_5.context, [[].concat([
                this._appEl_1,
                this._text_2
            ])], null);
        this._expr_1 = import7.UNINITIALIZED;
        var disposable_0 = this.renderer.listen(this._el_0, 'click', this.eventHandler(this._handle_click_0_0.bind(this)));
        this._expr_5 = import7.UNINITIALIZED;
        this.init([].concat([this._el_0]), [
            this._el_0,
            this._anchor_1,
            this._text_2
        ], [disposable_0], []);
        return null;
    };
    _View_ActionSheetCmp3.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import19.TemplateRef) && (1 === requestNodeIndex))) {
            return this._TemplateRef_1_5;
        }
        if (((token === import22.NgIf) && (1 === requestNodeIndex))) {
            return this._NgIf_1_6.context;
        }
        if (((token === import27.NgClass) && ((0 <= requestNodeIndex) && (requestNodeIndex <= 2)))) {
            return this._NgClass_0_4.context;
        }
        if (((token === import28.Button) && ((0 <= requestNodeIndex) && (requestNodeIndex <= 2)))) {
            return this._Button_0_5.context;
        }
        return notFoundResult;
    };
    _View_ActionSheetCmp3.prototype.detectChangesInternal = function (throwOnChange) {
        var currVal_2 = 'disable-hover';
        this._NgClass_0_4.check_klass(currVal_2, throwOnChange, false);
        var currVal_3 = this.context.$implicit.cssClass;
        this._NgClass_0_4.check_ngClass(currVal_3, throwOnChange, false);
        this._NgClass_0_4.detectChangesInternal(this, this._el_0, throwOnChange);
        if (this._Button_0_5.detectChangesInternal(this, this._el_0, throwOnChange)) {
            this._appEl_0.componentView.markAsCheckOnce();
        }
        var currVal_4 = this.context.$implicit.icon;
        this._NgIf_1_6.check_ngIf(currVal_4, throwOnChange, false);
        this._NgIf_1_6.detectChangesInternal(this, this._anchor_1, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        if (!throwOnChange) {
            if ((this.numberOfChecks === 0)) {
                this._Button_0_5.context.ngAfterContentInit();
            }
        }
        var currVal_1 = (this.context.$implicit.icon ? '' : null);
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setElementAttribute(this._el_0, 'icon-left', ((currVal_1 == null) ? null : currVal_1.toString()));
            this._expr_1 = currVal_1;
        }
        var currVal_5 = import4.interpolate(1, '', this.context.$implicit.text, '');
        if (import4.checkBinding(throwOnChange, this._expr_5, currVal_5)) {
            this.renderer.setText(this._text_2, currVal_5);
            this._expr_5 = currVal_5;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ActionSheetCmp3.prototype._handle_click_0_0 = function ($event) {
        this.markPathToRootAsCheckOnce();
        var pd_0 = (this.parent.context.click(this.context.$implicit) !== false);
        return (true && pd_0);
    };
    return _View_ActionSheetCmp3;
}(import1.AppView));
function viewFactory_ActionSheetCmp3(viewUtils, parentInjector, declarationEl) {
    return new _View_ActionSheetCmp3(viewUtils, parentInjector, declarationEl);
}
var _View_ActionSheetCmp4 = (function (_super) {
    __extends(_View_ActionSheetCmp4, _super);
    function _View_ActionSheetCmp4(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp4, renderType_ActionSheetCmp, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp4.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.renderer.createElement(null, 'ion-icon', null);
        this.renderer.setElementAttribute(this._el_0, 'class', 'action-sheet-icon');
        this.renderer.setElementAttribute(this._el_0, 'role', 'img');
        this._Icon_0_3 = new import29.Wrapper_Icon(this.parent.parent.parentInjector.get(import9.Config), new import10.ElementRef(this._el_0), this.renderer);
        this._expr_1 = import7.UNINITIALIZED;
        this.init([].concat([this._el_0]), [this._el_0], [], []);
        return null;
    };
    _View_ActionSheetCmp4.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import30.Icon) && (0 === requestNodeIndex))) {
            return this._Icon_0_3.context;
        }
        return notFoundResult;
    };
    _View_ActionSheetCmp4.prototype.detectChangesInternal = function (throwOnChange) {
        var currVal_0 = this.parent.context.$implicit.icon;
        this._Icon_0_3.check_name(currVal_0, throwOnChange, false);
        this._Icon_0_3.detectChangesInternal(this, this._el_0, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        var currVal_1 = this._Icon_0_3.context._hidden;
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setElementClass(this._el_0, 'hide', currVal_1);
            this._expr_1 = currVal_1;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ActionSheetCmp4.prototype.destroyInternal = function () {
        this._Icon_0_3.context.ngOnDestroy();
    };
    return _View_ActionSheetCmp4;
}(import1.AppView));
function viewFactory_ActionSheetCmp4(viewUtils, parentInjector, declarationEl) {
    return new _View_ActionSheetCmp4(viewUtils, parentInjector, declarationEl);
}
var _View_ActionSheetCmp5 = (function (_super) {
    __extends(_View_ActionSheetCmp5, _super);
    function _View_ActionSheetCmp5(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp5, renderType_ActionSheetCmp, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp5.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.renderer.createElement(null, 'div', null);
        this.renderer.setElementAttribute(this._el_0, 'class', 'action-sheet-group');
        this._el_1 = this.renderer.createElement(this._el_0, 'button', null);
        this.renderer.setElementAttribute(this._el_1, 'class', 'action-sheet-cancel disable-hover');
        this.renderer.setElementAttribute(this._el_1, 'ion-button', 'action-sheet-button');
        this._appEl_1 = new import3.AppElement(1, 0, this, this._el_1);
        var compView_1 = import25.viewFactory_Button0(this.viewUtils, this.injector(1), this._appEl_1);
        this._NgClass_1_4 = new import24.Wrapper_NgClass(this.parent.parentInjector.get(import20.IterableDiffers), this.parent.parentInjector.get(import26.KeyValueDiffers), new import10.ElementRef(this._el_1), this.renderer);
        this._Button_1_5 = new import25.Wrapper_Button(null, 'action-sheet-button', this.parent.parentInjector.get(import9.Config), new import10.ElementRef(this._el_1), this.renderer);
        this._appEl_1.initComponent(this._Button_1_5.context, [], compView_1);
        this._anchor_2 = this.renderer.createTemplateAnchor(null, null);
        this._appEl_2 = new import3.AppElement(2, 1, this, this._anchor_2);
        this._TemplateRef_2_5 = new import19.TemplateRef_(this._appEl_2, viewFactory_ActionSheetCmp6);
        this._NgIf_2_6 = new import16.Wrapper_NgIf(this._appEl_2.vcRef, this._TemplateRef_2_5);
        this._text_3 = this.renderer.createText(null, '', null);
        compView_1.create(this._Button_1_5.context, [[].concat([
                this._appEl_2,
                this._text_3
            ])], null);
        this._expr_1 = import7.UNINITIALIZED;
        var disposable_0 = this.renderer.listen(this._el_1, 'click', this.eventHandler(this._handle_click_1_0.bind(this)));
        this._expr_5 = import7.UNINITIALIZED;
        this.init([].concat([this._el_0]), [
            this._el_0,
            this._el_1,
            this._anchor_2,
            this._text_3
        ], [disposable_0], []);
        return null;
    };
    _View_ActionSheetCmp5.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import19.TemplateRef) && (2 === requestNodeIndex))) {
            return this._TemplateRef_2_5;
        }
        if (((token === import22.NgIf) && (2 === requestNodeIndex))) {
            return this._NgIf_2_6.context;
        }
        if (((token === import27.NgClass) && ((1 <= requestNodeIndex) && (requestNodeIndex <= 3)))) {
            return this._NgClass_1_4.context;
        }
        if (((token === import28.Button) && ((1 <= requestNodeIndex) && (requestNodeIndex <= 3)))) {
            return this._Button_1_5.context;
        }
        return notFoundResult;
    };
    _View_ActionSheetCmp5.prototype.detectChangesInternal = function (throwOnChange) {
        var currVal_2 = 'action-sheet-cancel disable-hover';
        this._NgClass_1_4.check_klass(currVal_2, throwOnChange, false);
        var currVal_3 = this.parent.context.d.cancelButton.cssClass;
        this._NgClass_1_4.check_ngClass(currVal_3, throwOnChange, false);
        this._NgClass_1_4.detectChangesInternal(this, this._el_1, throwOnChange);
        if (this._Button_1_5.detectChangesInternal(this, this._el_1, throwOnChange)) {
            this._appEl_1.componentView.markAsCheckOnce();
        }
        var currVal_4 = this.parent.context.d.cancelButton.icon;
        this._NgIf_2_6.check_ngIf(currVal_4, throwOnChange, false);
        this._NgIf_2_6.detectChangesInternal(this, this._anchor_2, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        if (!throwOnChange) {
            if ((this.numberOfChecks === 0)) {
                this._Button_1_5.context.ngAfterContentInit();
            }
        }
        var currVal_1 = (this.parent.context.d.cancelButton.icon ? '' : null);
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setElementAttribute(this._el_1, 'icon-left', ((currVal_1 == null) ? null : currVal_1.toString()));
            this._expr_1 = currVal_1;
        }
        var currVal_5 = import4.interpolate(1, '', this.parent.context.d.cancelButton.text, '');
        if (import4.checkBinding(throwOnChange, this._expr_5, currVal_5)) {
            this.renderer.setText(this._text_3, currVal_5);
            this._expr_5 = currVal_5;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ActionSheetCmp5.prototype._handle_click_1_0 = function ($event) {
        this.markPathToRootAsCheckOnce();
        var pd_0 = (this.parent.context.click(this.parent.context.d.cancelButton) !== false);
        return (true && pd_0);
    };
    return _View_ActionSheetCmp5;
}(import1.AppView));
function viewFactory_ActionSheetCmp5(viewUtils, parentInjector, declarationEl) {
    return new _View_ActionSheetCmp5(viewUtils, parentInjector, declarationEl);
}
var _View_ActionSheetCmp6 = (function (_super) {
    __extends(_View_ActionSheetCmp6, _super);
    function _View_ActionSheetCmp6(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ActionSheetCmp6, renderType_ActionSheetCmp, import6.ViewType.EMBEDDED, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ActionSheetCmp6.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.renderer.createElement(null, 'ion-icon', null);
        this.renderer.setElementAttribute(this._el_0, 'class', 'action-sheet-icon');
        this.renderer.setElementAttribute(this._el_0, 'role', 'img');
        this._Icon_0_3 = new import29.Wrapper_Icon(this.parent.parent.parentInjector.get(import9.Config), new import10.ElementRef(this._el_0), this.renderer);
        this._expr_1 = import7.UNINITIALIZED;
        this.init([].concat([this._el_0]), [this._el_0], [], []);
        return null;
    };
    _View_ActionSheetCmp6.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import30.Icon) && (0 === requestNodeIndex))) {
            return this._Icon_0_3.context;
        }
        return notFoundResult;
    };
    _View_ActionSheetCmp6.prototype.detectChangesInternal = function (throwOnChange) {
        var currVal_0 = this.parent.parent.context.d.cancelButton.icon;
        this._Icon_0_3.check_name(currVal_0, throwOnChange, false);
        this._Icon_0_3.detectChangesInternal(this, this._el_0, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        var currVal_1 = this._Icon_0_3.context._hidden;
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setElementClass(this._el_0, 'hide', currVal_1);
            this._expr_1 = currVal_1;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ActionSheetCmp6.prototype.destroyInternal = function () {
        this._Icon_0_3.context.ngOnDestroy();
    };
    return _View_ActionSheetCmp6;
}(import1.AppView));
function viewFactory_ActionSheetCmp6(viewUtils, parentInjector, declarationEl) {
    return new _View_ActionSheetCmp6(viewUtils, parentInjector, declarationEl);
}
