// ionic-angular/components/virtual-scroll/virtual-item.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from 'ionic-angular/components/virtual-scroll/virtual-item';
import * as import1 from '@angular/core/src/linker/view';
export class Wrapper_VirtualHeader {
  context:import0.VirtualHeader;
  changed:boolean;
  constructor(p0:any) {
    this.changed = false;
    this.context = new import0.VirtualHeader(p0);
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
export class Wrapper_VirtualFooter {
  context:import0.VirtualFooter;
  changed:boolean;
  constructor(p0:any) {
    this.changed = false;
    this.context = new import0.VirtualFooter(p0);
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
export class Wrapper_VirtualItem {
  context:import0.VirtualItem;
  changed:boolean;
  constructor(p0:any,p1:any) {
    this.changed = false;
    this.context = new import0.VirtualItem(p0,p1);
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
