// @ionic/app-scripts/dist/dev-server/dev-server.js
// Ionic Dev Server: Server Side Logger
"use strict";
var logger_1 = require('../util/logger');
var config_1 = require('../util/config');
var events_1 = require('../util/events');
var ws_1 = require('ws');
var wsServer;
var msgToClient = [];
function createDevServer() {
    events_1.on(events_1.EventType.TaskEvent, function (taskEvent) {
        var msg = {
            category: 'taskEvent',
            type: taskEvent.scope,
            data: taskEvent
        };
        queueMessageSend(msg);
    });
    events_1.on(events_1.EventType.TranspileDiagnostics, function (diagnostic) {
        var msg = {
            category: 'transpileDiagnostics',
            type: 'typescript',
            data: diagnostic
        };
        queueMessageSend(msg);
    });
    // create web socket server
    var wss = new ws_1.Server({ port: getWsPort() });
    wss.on('connection', function (ws) {
        wsServer = ws;
        wsServer.on('message', function (incomingMessage) {
            // incoming message from the client
            try {
                printClientMessage(JSON.parse(incomingMessage));
            }
            catch (e) {
                logger_1.Logger.error("error opening ws message: " + incomingMessage);
            }
        });
        drainMessageQueue();
    });
}
exports.createDevServer = createDevServer;
function queueMessageSend(msg) {
    msgToClient.push(msg);
    drainMessageQueue();
}
function drainMessageQueue() {
    if (wsServer) {
        var msg;
        while (msg = msgToClient.shift()) {
            try {
                wsServer.send(JSON.stringify(msg));
            }
            catch (e) {
                logger_1.Logger.error("error sending client ws, " + e);
            }
        }
    }
}
function printClientMessage(msg) {
    if (msg.data) {
        switch (msg.category) {
            case 'console':
                printConsole(msg);
                break;
            case 'exception':
                printException(msg);
                break;
        }
    }
}
function printConsole(msg) {
    var args = msg.data;
    args[0] = "console." + msg.type + ": " + args[0];
    switch (msg.type) {
        case 'error':
            logger_1.Logger.error.apply(this, args);
            break;
        case 'warn':
            logger_1.Logger.warn.apply(this, args);
            break;
        case 'debug':
            logger_1.Logger.debug.apply(this, args);
            break;
        default:
            logger_1.Logger.info.apply(this, args);
            break;
    }
}
function printException(msg) {
}
function sendClientConsoleLogs() {
    return config_1.hasConfigValue('--consolelogs', '-c', 'ionic_consolelogs', false);
}
exports.sendClientConsoleLogs = sendClientConsoleLogs;
function getWsPort() {
    var port = config_1.getConfigValueDefault('--dev-logger-port', null, 'ionic_dev_logger_port', null);
    if (port) {
        return parseInt(port, 10);
    }
    return DEV_LOGGER_DEFAULT_PORT;
}
exports.getWsPort = getWsPort;
var DEV_LOGGER_DEFAULT_PORT = 53703;
