// @angular/forms/src/directives/select_control_value_accessor.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from '@angular/forms/src/directives/select_control_value_accessor';
import * as import1 from '@angular/core/src/linker/view';
import * as import2 from '@angular/core/src/change_detection/change_detection';
import * as import3 from '@angular/core/src/linker/view_utils';
export class Wrapper_SelectControlValueAccessor {
  context:import0.SelectControlValueAccessor;
  changed:boolean;
  constructor(p0:any,p1:any) {
    this.changed = false;
    this.context = new import0.SelectControlValueAccessor(p0,p1);
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
export class Wrapper_NgSelectOption {
  context:import0.NgSelectOption;
  changed:boolean;
  /*private*/ _ngValue:any;
  /*private*/ _value:any;
  constructor(p0:any,p1:any,p2:any) {
    this.changed = false;
    this._ngValue = import2.UNINITIALIZED;
    this._value = import2.UNINITIALIZED;
    this.context = new import0.NgSelectOption(p0,p1,p2);
  }
  check_ngValue(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import3.checkBinding(throwOnChange,this._ngValue,currValue))) {
      this.changed = true;
      this.context.ngValue = currValue;
      this._ngValue = currValue;
    }
  }
  check_value(currValue:any,throwOnChange:boolean,forceUpdate:boolean):void {
    if ((forceUpdate || import3.checkBinding(throwOnChange,this._value,currValue))) {
      this.changed = true;
      this.context.value = currValue;
      this._value = currValue;
    }
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
