// ionic-angular/components/modal/modal-component.ngfactory.js
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import * as import0 from 'ionic-angular/components/modal/modal-component';
import * as import1 from '@angular/core/src/linker/view';
import * as import3 from '@angular/core/src/linker/element';
import * as import6 from '@angular/core/src/linker/view_type';
import * as import7 from '@angular/core/src/change_detection/change_detection';
import * as import8 from '@angular/core/src/linker/component_factory_resolver';
import * as import9 from 'ionic-angular/navigation/nav-params';
import * as import10 from 'ionic-angular/navigation/view-controller';
import * as import11 from '@angular/core/src/metadata/view';
import * as import12 from '@angular/core/src/linker/component_factory';
import * as import13 from '@angular/core/src/linker/query_list';
import * as import14 from '../backdrop/backdrop.ngfactory';
import * as import15 from 'ionic-angular/gestures/gesture-controller';
import * as import16 from '@angular/core/src/linker/element_ref';
import * as import17 from 'ionic-angular/components/backdrop/backdrop';
export var Wrapper_ModalCmp = (function () {
    function Wrapper_ModalCmp(p0, p1, p2, p3) {
        this.changed = false;
        this.context = new import0.ModalCmp(p0, p1, p2, p3);
    }
    Wrapper_ModalCmp.prototype.detectChangesInternal = function (view, el, throwOnChange) {
        var changed = this.changed;
        this.changed = false;
        return changed;
    };
    return Wrapper_ModalCmp;
}());
var renderType_ModalCmp_Host = null;
var _View_ModalCmp_Host0 = (function (_super) {
    __extends(_View_ModalCmp_Host0, _super);
    function _View_ModalCmp_Host0(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ModalCmp_Host0, renderType_ModalCmp_Host, import6.ViewType.HOST, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ModalCmp_Host0.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.selectOrCreateHostElement('ion-modal', rootSelector, null);
        this._appEl_0 = new import3.AppElement(0, null, this, this._el_0);
        var compView_0 = viewFactory_ModalCmp0(this.viewUtils, this.injector(0), this._appEl_0);
        this._ModalCmp_0_4 = new Wrapper_ModalCmp(this.parentInjector.get(import8.ComponentFactoryResolver), this.renderer, this.parentInjector.get(import9.NavParams), this.parentInjector.get(import10.ViewController));
        this._appEl_0.initComponent(this._ModalCmp_0_4.context, [], compView_0);
        compView_0.create(this._ModalCmp_0_4.context, this.projectableNodes, null);
        var disposable_0 = this.renderer.listenGlobal('body', 'keyup', this.eventHandler(this._handle_keyup_0_0.bind(this)));
        this.init([].concat([this._el_0]), [this._el_0], [disposable_0], []);
        return this._appEl_0;
    };
    _View_ModalCmp_Host0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import0.ModalCmp) && (0 === requestNodeIndex))) {
            return this._ModalCmp_0_4.context;
        }
        return notFoundResult;
    };
    _View_ModalCmp_Host0.prototype.detectChangesInternal = function (throwOnChange) {
        this._ModalCmp_0_4.detectChangesInternal(this, this._el_0, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ModalCmp_Host0.prototype._handle_keyup_0_0 = function ($event) {
        this._appEl_0.componentView.markPathToRootAsCheckOnce();
        var pd_0 = (this._ModalCmp_0_4.context._keyUp($event) !== false);
        return (true && pd_0);
    };
    return _View_ModalCmp_Host0;
}(import1.AppView));
function viewFactory_ModalCmp_Host0(viewUtils, parentInjector, declarationEl) {
    if ((renderType_ModalCmp_Host === null)) {
        (renderType_ModalCmp_Host = viewUtils.createRenderComponentType('', 0, import11.ViewEncapsulation.None, [], {}));
    }
    return new _View_ModalCmp_Host0(viewUtils, parentInjector, declarationEl);
}
export var ModalCmpNgFactory = new import12.ComponentFactory('ion-modal', viewFactory_ModalCmp_Host0, import0.ModalCmp);
var styles_ModalCmp = [];
var renderType_ModalCmp = null;
var _View_ModalCmp0 = (function (_super) {
    __extends(_View_ModalCmp0, _super);
    function _View_ModalCmp0(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_ModalCmp0, renderType_ModalCmp, import6.ViewType.COMPONENT, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_ModalCmp0.prototype.createInternal = function (rootSelector) {
        var parentRenderNode = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
        this._viewQuery_viewport_0 = new import13.QueryList();
        this._el_0 = this.renderer.createElement(parentRenderNode, 'ion-backdrop', null);
        this.renderer.setElementAttribute(this._el_0, 'disable-activated', '');
        this.renderer.setElementAttribute(this._el_0, 'disableScroll', 'false');
        this.renderer.setElementAttribute(this._el_0, 'role', 'presentation');
        this.renderer.setElementAttribute(this._el_0, 'tappable', '');
        this._Backdrop_0_3 = new import14.Wrapper_Backdrop(this.parentInjector.get(import15.GestureController), new import16.ElementRef(this._el_0), this.renderer);
        this._el_1 = this.renderer.createElement(parentRenderNode, 'div', null);
        this.renderer.setElementAttribute(this._el_1, 'class', 'modal-wrapper');
        this._el_2 = this.renderer.createElement(this._el_1, 'div', null);
        this.renderer.setElementAttribute(this._el_2, 'nav-viewport', '');
        this._appEl_2 = new import3.AppElement(2, 1, this, this._el_2);
        var disposable_0 = this.renderer.listen(this._el_0, 'click', this.eventHandler(this._handle_click_0_0.bind(this)));
        this._viewQuery_viewport_0.reset([this._appEl_2.vcRef]);
        this.context._viewport = this._viewQuery_viewport_0.first;
        this.init([], [
            this._el_0,
            this._el_1,
            this._el_2
        ], [disposable_0], []);
        return null;
    };
    _View_ModalCmp0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import17.Backdrop) && (0 === requestNodeIndex))) {
            return this._Backdrop_0_3.context;
        }
        return notFoundResult;
    };
    _View_ModalCmp0.prototype.detectChangesInternal = function (throwOnChange) {
        var currVal_1 = 'false';
        this._Backdrop_0_3.check_disableScroll(currVal_1, throwOnChange, false);
        this._Backdrop_0_3.detectChangesInternal(this, this._el_0, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_ModalCmp0.prototype.destroyInternal = function () {
        this._Backdrop_0_3.context.ngOnDestroy();
    };
    _View_ModalCmp0.prototype._handle_click_0_0 = function ($event) {
        this.markPathToRootAsCheckOnce();
        var pd_0 = (this.context._bdClick() !== false);
        return (true && pd_0);
    };
    return _View_ModalCmp0;
}(import1.AppView));
export function viewFactory_ModalCmp0(viewUtils, parentInjector, declarationEl) {
    if ((renderType_ModalCmp === null)) {
        (renderType_ModalCmp = viewUtils.createRenderComponentType('', 0, import11.ViewEncapsulation.None, styles_ModalCmp, {}));
    }
    return new _View_ModalCmp0(viewUtils, parentInjector, declarationEl);
}
