// ionic-angular/components/app/app-root.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from 'ionic-angular/components/app/app-root';
import * as import1 from '@angular/core/src/linker/view';
import * as import2 from '@angular/core/src/render/api';
import * as import3 from '@angular/core/src/linker/element';
import * as import4 from '@angular/core/src/linker/view_utils';
import * as import5 from '@angular/core/src/di/injector';
import * as import6 from '@angular/core/src/linker/view_type';
import * as import7 from '@angular/core/src/change_detection/change_detection';
import * as import8 from '@angular/core/src/linker/component_factory_resolver';
import * as import9 from '@angular/core/src/linker/element_ref';
import * as import10 from 'ionic-angular/config/config';
import * as import11 from 'ionic-angular/platform/platform';
import * as import12 from 'ionic-angular/components/app/app';
import * as import13 from '@angular/core/src/metadata/view';
import * as import14 from '@angular/core/src/linker/component_factory';
import * as import15 from '@angular/core/src/linker/query_list';
import * as import16 from '../nav/overlay-portal.ngfactory';
import * as import17 from 'ionic-angular/util/keyboard';
import * as import18 from '@angular/core/src/zone/ng_zone';
import * as import19 from 'ionic-angular/gestures/gesture-controller';
import * as import20 from 'ionic-angular/transitions/transition-controller';
import * as import21 from 'ionic-angular/navigation/deep-linker';
import * as import22 from 'ionic-angular/components/nav/overlay-portal';
export class Wrapper_IonicApp {
  context:import0.IonicApp;
  changed:boolean;
  constructor(p0:any,p1:any,p2:any,p3:any,p4:any,p5:any,p6:any) {
    this.changed = false;
    this.context = new import0.IonicApp(p0,p1,p2,p3,p4,p5,p6);
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    if (!throwOnChange) { if ((view.numberOfChecks === 0)) { this.context.ngOnInit(); } }
    return changed;
  }
}
var renderType_IonicApp_Host:import2.RenderComponentType = (null as any);
class _View_IonicApp_Host0 extends import1.AppView<any> {
  _el_0:any;
  /*private*/ _appEl_0:import3.AppElement;
  _IonicApp_0_4:Wrapper_IonicApp;
  constructor(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement) {
    super(_View_IonicApp_Host0,renderType_IonicApp_Host,import6.ViewType.HOST,viewUtils,parentInjector,declarationEl,import7.ChangeDetectorStatus.CheckAlways);
  }
  createInternal(rootSelector:string):import3.AppElement {
    this._el_0 = this.selectOrCreateHostElement('ion-app',rootSelector,(null as any));
    this._appEl_0 = new import3.AppElement(0,(null as any),this,this._el_0);
    var compView_0:any = viewFactory_IonicApp0(this.viewUtils,this.injector(0),this._appEl_0);
    this._IonicApp_0_4 = new Wrapper_IonicApp(this.parentInjector.get(import0.AppRootToken),this.parentInjector.get(import8.ComponentFactoryResolver),new import9.ElementRef(this._el_0),this.renderer,this.parentInjector.get(import10.Config),this.parentInjector.get(import11.Platform),this.parentInjector.get(import12.App));
    this._appEl_0.initComponent(this._IonicApp_0_4.context,([] as any[]),compView_0);
    compView_0.create(this._IonicApp_0_4.context,this.projectableNodes,(null as any));
    this.init(([] as any[]).concat([this._el_0]),[this._el_0],([] as any[]),([] as any[]));
    return this._appEl_0;
  }
  injectorGetInternal(token:any,requestNodeIndex:number,notFoundResult:any):any {
    if (((token === import0.IonicApp) && (0 === requestNodeIndex))) { return this._IonicApp_0_4.context; }
    return notFoundResult;
  }
  detectChangesInternal(throwOnChange:boolean):void {
    this._IonicApp_0_4.detectChangesInternal(this,this._el_0,throwOnChange);
    this.detectContentChildrenChanges(throwOnChange);
    this.detectViewChildrenChanges(throwOnChange);
  }
}
function viewFactory_IonicApp_Host0(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement):import1.AppView<any> {
  if ((renderType_IonicApp_Host === (null as any))) { (renderType_IonicApp_Host = viewUtils.createRenderComponentType('',0,import13.ViewEncapsulation.None,([] as any[]),{})); }
  return new _View_IonicApp_Host0(viewUtils,parentInjector,declarationEl);
}
export const IonicAppNgFactory:import14.ComponentFactory<import0.IonicApp> = new import14.ComponentFactory<import0.IonicApp>('ion-app',viewFactory_IonicApp_Host0,import0.IonicApp);
const styles_IonicApp:any[] = ([] as any[]);
var renderType_IonicApp:import2.RenderComponentType = (null as any);
class _View_IonicApp0 extends import1.AppView<import0.IonicApp> {
  _viewQuery_viewport_0:import15.QueryList<any>;
  _viewQuery_modalPortal_1:import15.QueryList<any>;
  _viewQuery_overlayPortal_2:import15.QueryList<any>;
  _viewQuery_loadingPortal_3:import15.QueryList<any>;
  _viewQuery_toastPortal_4:import15.QueryList<any>;
  _el_0:any;
  /*private*/ _appEl_0:import3.AppElement;
  _el_1:any;
  /*private*/ _appEl_1:import3.AppElement;
  _OverlayPortal_1_5:import16.Wrapper_OverlayPortal;
  _el_2:any;
  /*private*/ _appEl_2:import3.AppElement;
  _OverlayPortal_2_5:import16.Wrapper_OverlayPortal;
  _el_3:any;
  /*private*/ _appEl_3:import3.AppElement;
  _OverlayPortal_3_5:import16.Wrapper_OverlayPortal;
  _el_4:any;
  /*private*/ _appEl_4:import3.AppElement;
  _OverlayPortal_4_5:import16.Wrapper_OverlayPortal;
  _el_5:any;
  constructor(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement) {
    super(_View_IonicApp0,renderType_IonicApp,import6.ViewType.COMPONENT,viewUtils,parentInjector,declarationEl,import7.ChangeDetectorStatus.CheckAlways);
  }
  createInternal(rootSelector:string):import3.AppElement {
    const parentRenderNode:any = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
    this._viewQuery_viewport_0 = new import15.QueryList<any>();
    this._viewQuery_modalPortal_1 = new import15.QueryList<any>();
    this._viewQuery_overlayPortal_2 = new import15.QueryList<any>();
    this._viewQuery_loadingPortal_3 = new import15.QueryList<any>();
    this._viewQuery_toastPortal_4 = new import15.QueryList<any>();
    this._el_0 = this.renderer.createElement(parentRenderNode,'div',(null as any));
    this.renderer.setElementAttribute(this._el_0,'app-viewport','');
    this._appEl_0 = new import3.AppElement(0,(null as any),this,this._el_0);
    this._el_1 = this.renderer.createElement(parentRenderNode,'div',(null as any));
    this.renderer.setElementAttribute(this._el_1,'overlay-portal','');
    this._appEl_1 = new import3.AppElement(1,(null as any),this,this._el_1);
    this._OverlayPortal_1_5 = new import16.Wrapper_OverlayPortal(this.parentInjector.get(import12.App),this.parentInjector.get(import10.Config),this.parentInjector.get(import17.Keyboard),new import9.ElementRef(this._el_1),this.parentInjector.get(import18.NgZone),this.renderer,this.parentInjector.get(import8.ComponentFactoryResolver),this.parentInjector.get(import19.GestureController),this.parentInjector.get(import20.TransitionController),this.parentInjector.get(import21.DeepLinker,(null as any)),this._appEl_1.vcRef);
    this._el_2 = this.renderer.createElement(parentRenderNode,'div',(null as any));
    this.renderer.setElementAttribute(this._el_2,'overlay-portal','');
    this._appEl_2 = new import3.AppElement(2,(null as any),this,this._el_2);
    this._OverlayPortal_2_5 = new import16.Wrapper_OverlayPortal(this.parentInjector.get(import12.App),this.parentInjector.get(import10.Config),this.parentInjector.get(import17.Keyboard),new import9.ElementRef(this._el_2),this.parentInjector.get(import18.NgZone),this.renderer,this.parentInjector.get(import8.ComponentFactoryResolver),this.parentInjector.get(import19.GestureController),this.parentInjector.get(import20.TransitionController),this.parentInjector.get(import21.DeepLinker,(null as any)),this._appEl_2.vcRef);
    this._el_3 = this.renderer.createElement(parentRenderNode,'div',(null as any));
    this.renderer.setElementAttribute(this._el_3,'class','loading-portal');
    this.renderer.setElementAttribute(this._el_3,'overlay-portal','');
    this._appEl_3 = new import3.AppElement(3,(null as any),this,this._el_3);
    this._OverlayPortal_3_5 = new import16.Wrapper_OverlayPortal(this.parentInjector.get(import12.App),this.parentInjector.get(import10.Config),this.parentInjector.get(import17.Keyboard),new import9.ElementRef(this._el_3),this.parentInjector.get(import18.NgZone),this.renderer,this.parentInjector.get(import8.ComponentFactoryResolver),this.parentInjector.get(import19.GestureController),this.parentInjector.get(import20.TransitionController),this.parentInjector.get(import21.DeepLinker,(null as any)),this._appEl_3.vcRef);
    this._el_4 = this.renderer.createElement(parentRenderNode,'div',(null as any));
    this.renderer.setElementAttribute(this._el_4,'class','toast-portal');
    this.renderer.setElementAttribute(this._el_4,'overlay-portal','');
    this._appEl_4 = new import3.AppElement(4,(null as any),this,this._el_4);
    this._OverlayPortal_4_5 = new import16.Wrapper_OverlayPortal(this.parentInjector.get(import12.App),this.parentInjector.get(import10.Config),this.parentInjector.get(import17.Keyboard),new import9.ElementRef(this._el_4),this.parentInjector.get(import18.NgZone),this.renderer,this.parentInjector.get(import8.ComponentFactoryResolver),this.parentInjector.get(import19.GestureController),this.parentInjector.get(import20.TransitionController),this.parentInjector.get(import21.DeepLinker,(null as any)),this._appEl_4.vcRef);
    this._el_5 = this.renderer.createElement(parentRenderNode,'div',(null as any));
    this.renderer.setElementAttribute(this._el_5,'class','click-block');
    this._viewQuery_viewport_0.reset([this._appEl_0.vcRef]);
    this.context._viewport = this._viewQuery_viewport_0.first;
    this._viewQuery_modalPortal_1.reset([this._OverlayPortal_1_5.context]);
    this.context._modalPortal = this._viewQuery_modalPortal_1.first;
    this._viewQuery_overlayPortal_2.reset([this._OverlayPortal_2_5.context]);
    this.context._overlayPortal = this._viewQuery_overlayPortal_2.first;
    this._viewQuery_loadingPortal_3.reset([this._OverlayPortal_3_5.context]);
    this.context._loadingPortal = this._viewQuery_loadingPortal_3.first;
    this._viewQuery_toastPortal_4.reset([this._OverlayPortal_4_5.context]);
    this.context._toastPortal = this._viewQuery_toastPortal_4.first;
    this.init(([] as any[]),[
      this._el_0,
      this._el_1,
      this._el_2,
      this._el_3,
      this._el_4,
      this._el_5
    ]
    ,([] as any[]),([] as any[]));
    return (null as any);
  }
  injectorGetInternal(token:any,requestNodeIndex:number,notFoundResult:any):any {
    if (((token === import22.OverlayPortal) && (1 === requestNodeIndex))) { return this._OverlayPortal_1_5.context; }
    if (((token === import22.OverlayPortal) && (2 === requestNodeIndex))) { return this._OverlayPortal_2_5.context; }
    if (((token === import22.OverlayPortal) && (3 === requestNodeIndex))) { return this._OverlayPortal_3_5.context; }
    if (((token === import22.OverlayPortal) && (4 === requestNodeIndex))) { return this._OverlayPortal_4_5.context; }
    return notFoundResult;
  }
  detectChangesInternal(throwOnChange:boolean):void {
    this._OverlayPortal_1_5.detectChangesInternal(this,this._el_1,throwOnChange);
    this._OverlayPortal_2_5.detectChangesInternal(this,this._el_2,throwOnChange);
    this._OverlayPortal_3_5.detectChangesInternal(this,this._el_3,throwOnChange);
    this._OverlayPortal_4_5.detectChangesInternal(this,this._el_4,throwOnChange);
    this.detectContentChildrenChanges(throwOnChange);
    this.detectViewChildrenChanges(throwOnChange);
  }
}
export function viewFactory_IonicApp0(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement):import1.AppView<import0.IonicApp> {
  if ((renderType_IonicApp === (null as any))) { (renderType_IonicApp = viewUtils.createRenderComponentType('',0,import13.ViewEncapsulation.None,styles_IonicApp,{})); }
  return new _View_IonicApp0(viewUtils,parentInjector,declarationEl);
}
