// ionic-angular/components/popover/popover-component.ngfactory.ts
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
 /* tslint:disable */

import * as import0 from 'ionic-angular/components/popover/popover-component';
import * as import1 from '@angular/core/src/linker/view';
import * as import2 from '@angular/core/src/render/api';
import * as import3 from '@angular/core/src/linker/element';
import * as import4 from '@angular/core/src/linker/view_utils';
import * as import5 from '@angular/core/src/di/injector';
import * as import6 from '@angular/core/src/linker/view_type';
import * as import7 from '@angular/core/src/change_detection/change_detection';
import * as import8 from '@angular/core/src/linker/component_factory_resolver';
import * as import9 from '@angular/core/src/linker/element_ref';
import * as import10 from 'ionic-angular/config/config';
import * as import11 from 'ionic-angular/navigation/nav-params';
import * as import12 from 'ionic-angular/navigation/view-controller';
import * as import13 from '@angular/core/src/metadata/view';
import * as import14 from '@angular/core/src/linker/component_factory';
import * as import15 from '@angular/core/src/linker/query_list';
import * as import16 from '../backdrop/backdrop.ngfactory';
import * as import17 from 'ionic-angular/gestures/gesture-controller';
import * as import18 from 'ionic-angular/components/backdrop/backdrop';
export class Wrapper_PopoverCmp {
  context:import0.PopoverCmp;
  changed:boolean;
  constructor(p0:any,p1:any,p2:any,p3:any,p4:any,p5:any) {
    this.changed = false;
    this.context = new import0.PopoverCmp(p0,p1,p2,p3,p4,p5);
  }
  detectChangesInternal(view:import1.AppView<any>,el:any,throwOnChange:boolean):boolean {
    var changed:any = this.changed;
    this.changed = false;
    return changed;
  }
}
var renderType_PopoverCmp_Host:import2.RenderComponentType = (null as any);
class _View_PopoverCmp_Host0 extends import1.AppView<any> {
  _el_0:any;
  /*private*/ _appEl_0:import3.AppElement;
  _PopoverCmp_0_4:Wrapper_PopoverCmp;
  constructor(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement) {
    super(_View_PopoverCmp_Host0,renderType_PopoverCmp_Host,import6.ViewType.HOST,viewUtils,parentInjector,declarationEl,import7.ChangeDetectorStatus.CheckAlways);
  }
  createInternal(rootSelector:string):import3.AppElement {
    this._el_0 = this.selectOrCreateHostElement('ion-popover',rootSelector,(null as any));
    this._appEl_0 = new import3.AppElement(0,(null as any),this,this._el_0);
    var compView_0:any = viewFactory_PopoverCmp0(this.viewUtils,this.injector(0),this._appEl_0);
    this._PopoverCmp_0_4 = new Wrapper_PopoverCmp(this.parentInjector.get(import8.ComponentFactoryResolver),new import9.ElementRef(this._el_0),this.renderer,this.parentInjector.get(import10.Config),this.parentInjector.get(import11.NavParams),this.parentInjector.get(import12.ViewController));
    this._appEl_0.initComponent(this._PopoverCmp_0_4.context,([] as any[]),compView_0);
    compView_0.create(this._PopoverCmp_0_4.context,this.projectableNodes,(null as any));
    var disposable_0:Function = this.renderer.listenGlobal('body','keyup',this.eventHandler(this._handle_keyup_0_0.bind(this)));
    this.init(([] as any[]).concat([this._el_0]),[this._el_0],[disposable_0],([] as any[]));
    return this._appEl_0;
  }
  injectorGetInternal(token:any,requestNodeIndex:number,notFoundResult:any):any {
    if (((token === import0.PopoverCmp) && (0 === requestNodeIndex))) { return this._PopoverCmp_0_4.context; }
    return notFoundResult;
  }
  detectChangesInternal(throwOnChange:boolean):void {
    this._PopoverCmp_0_4.detectChangesInternal(this,this._el_0,throwOnChange);
    this.detectContentChildrenChanges(throwOnChange);
    this.detectViewChildrenChanges(throwOnChange);
  }
  private _handle_keyup_0_0($event:any):boolean {
    this._appEl_0.componentView.markPathToRootAsCheckOnce();
    const pd_0:any = ((<any>this._PopoverCmp_0_4.context._keyUp($event)) !== false);
    return (true && pd_0);
  }
}
function viewFactory_PopoverCmp_Host0(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement):import1.AppView<any> {
  if ((renderType_PopoverCmp_Host === (null as any))) { (renderType_PopoverCmp_Host = viewUtils.createRenderComponentType('',0,import13.ViewEncapsulation.None,([] as any[]),{})); }
  return new _View_PopoverCmp_Host0(viewUtils,parentInjector,declarationEl);
}
export const PopoverCmpNgFactory:import14.ComponentFactory<import0.PopoverCmp> = new import14.ComponentFactory<import0.PopoverCmp>('ion-popover',viewFactory_PopoverCmp_Host0,import0.PopoverCmp);
const styles_PopoverCmp:any[] = ([] as any[]);
var renderType_PopoverCmp:import2.RenderComponentType = (null as any);
class _View_PopoverCmp0 extends import1.AppView<import0.PopoverCmp> {
  _viewQuery_viewport_0:import15.QueryList<any>;
  _el_0:any;
  _Backdrop_0_3:import16.Wrapper_Backdrop;
  _el_1:any;
  _el_2:any;
  _el_3:any;
  _el_4:any;
  _el_5:any;
  /*private*/ _appEl_5:import3.AppElement;
  /*private*/ _expr_1:any;
  constructor(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement) {
    super(_View_PopoverCmp0,renderType_PopoverCmp,import6.ViewType.COMPONENT,viewUtils,parentInjector,declarationEl,import7.ChangeDetectorStatus.CheckAlways);
  }
  createInternal(rootSelector:string):import3.AppElement {
    const parentRenderNode:any = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
    this._viewQuery_viewport_0 = new import15.QueryList<any>();
    this._el_0 = this.renderer.createElement(parentRenderNode,'ion-backdrop',(null as any));
    this.renderer.setElementAttribute(this._el_0,'disable-activated','');
    this.renderer.setElementAttribute(this._el_0,'role','presentation');
    this.renderer.setElementAttribute(this._el_0,'tappable','');
    this._Backdrop_0_3 = new import16.Wrapper_Backdrop(this.parentInjector.get(import17.GestureController),new import9.ElementRef(this._el_0),this.renderer);
    this._el_1 = this.renderer.createElement(parentRenderNode,'div',(null as any));
    this.renderer.setElementAttribute(this._el_1,'class','popover-wrapper');
    this._el_2 = this.renderer.createElement(this._el_1,'div',(null as any));
    this.renderer.setElementAttribute(this._el_2,'class','popover-arrow');
    this._el_3 = this.renderer.createElement(this._el_1,'div',(null as any));
    this.renderer.setElementAttribute(this._el_3,'class','popover-content');
    this._el_4 = this.renderer.createElement(this._el_3,'div',(null as any));
    this.renderer.setElementAttribute(this._el_4,'class','popover-viewport');
    this._el_5 = this.renderer.createElement(this._el_4,'div',(null as any));
    this.renderer.setElementAttribute(this._el_5,'nav-viewport','');
    this._appEl_5 = new import3.AppElement(5,4,this,this._el_5);
    this._expr_1 = import7.UNINITIALIZED;
    var disposable_0:Function = this.renderer.listen(this._el_0,'click',this.eventHandler(this._handle_click_0_0.bind(this)));
    this._viewQuery_viewport_0.reset([this._appEl_5.vcRef]);
    this.context._viewport = this._viewQuery_viewport_0.first;
    this.init(([] as any[]),[
      this._el_0,
      this._el_1,
      this._el_2,
      this._el_3,
      this._el_4,
      this._el_5
    ]
    ,[disposable_0],([] as any[]));
    return (null as any);
  }
  injectorGetInternal(token:any,requestNodeIndex:number,notFoundResult:any):any {
    if (((token === import18.Backdrop) && (0 === requestNodeIndex))) { return this._Backdrop_0_3.context; }
    return notFoundResult;
  }
  detectChangesInternal(throwOnChange:boolean):void {
    this._Backdrop_0_3.detectChangesInternal(this,this._el_0,throwOnChange);
    this.detectContentChildrenChanges(throwOnChange);
    const currVal_1:boolean = !this.context.d.showBackdrop;
    if (import4.checkBinding(throwOnChange,this._expr_1,currVal_1)) {
      this.renderer.setElementClass(this._el_0,'hide-backdrop',currVal_1);
      this._expr_1 = currVal_1;
    }
    this.detectViewChildrenChanges(throwOnChange);
  }
  destroyInternal():void {
    this._Backdrop_0_3.context.ngOnDestroy();
  }
  private _handle_click_0_0($event:any):boolean {
    this.markPathToRootAsCheckOnce();
    const pd_0:any = ((<any>this.context._bdClick()) !== false);
    return (true && pd_0);
  }
}
export function viewFactory_PopoverCmp0(viewUtils:import4.ViewUtils,parentInjector:import5.Injector,declarationEl:import3.AppElement):import1.AppView<import0.PopoverCmp> {
  if ((renderType_PopoverCmp === (null as any))) { (renderType_PopoverCmp = viewUtils.createRenderComponentType('',0,import13.ViewEncapsulation.None,styles_PopoverCmp,{})); }
  return new _View_PopoverCmp0(viewUtils,parentInjector,declarationEl);
}
