// ionic-angular/components/popover/popover-component.ngfactory.js
/**
 * This file is generated by the Angular 2 template compiler.
 * Do not edit.
 */
/* tslint:disable */
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
import * as import0 from 'ionic-angular/components/popover/popover-component';
import * as import1 from '@angular/core/src/linker/view';
import * as import3 from '@angular/core/src/linker/element';
import * as import4 from '@angular/core/src/linker/view_utils';
import * as import6 from '@angular/core/src/linker/view_type';
import * as import7 from '@angular/core/src/change_detection/change_detection';
import * as import8 from '@angular/core/src/linker/component_factory_resolver';
import * as import9 from '@angular/core/src/linker/element_ref';
import * as import10 from 'ionic-angular/config/config';
import * as import11 from 'ionic-angular/navigation/nav-params';
import * as import12 from 'ionic-angular/navigation/view-controller';
import * as import13 from '@angular/core/src/metadata/view';
import * as import14 from '@angular/core/src/linker/component_factory';
import * as import15 from '@angular/core/src/linker/query_list';
import * as import16 from '../backdrop/backdrop.ngfactory';
import * as import17 from 'ionic-angular/gestures/gesture-controller';
import * as import18 from 'ionic-angular/components/backdrop/backdrop';
export var Wrapper_PopoverCmp = (function () {
    function Wrapper_PopoverCmp(p0, p1, p2, p3, p4, p5) {
        this.changed = false;
        this.context = new import0.PopoverCmp(p0, p1, p2, p3, p4, p5);
    }
    Wrapper_PopoverCmp.prototype.detectChangesInternal = function (view, el, throwOnChange) {
        var changed = this.changed;
        this.changed = false;
        return changed;
    };
    return Wrapper_PopoverCmp;
}());
var renderType_PopoverCmp_Host = null;
var _View_PopoverCmp_Host0 = (function (_super) {
    __extends(_View_PopoverCmp_Host0, _super);
    function _View_PopoverCmp_Host0(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_PopoverCmp_Host0, renderType_PopoverCmp_Host, import6.ViewType.HOST, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_PopoverCmp_Host0.prototype.createInternal = function (rootSelector) {
        this._el_0 = this.selectOrCreateHostElement('ion-popover', rootSelector, null);
        this._appEl_0 = new import3.AppElement(0, null, this, this._el_0);
        var compView_0 = viewFactory_PopoverCmp0(this.viewUtils, this.injector(0), this._appEl_0);
        this._PopoverCmp_0_4 = new Wrapper_PopoverCmp(this.parentInjector.get(import8.ComponentFactoryResolver), new import9.ElementRef(this._el_0), this.renderer, this.parentInjector.get(import10.Config), this.parentInjector.get(import11.NavParams), this.parentInjector.get(import12.ViewController));
        this._appEl_0.initComponent(this._PopoverCmp_0_4.context, [], compView_0);
        compView_0.create(this._PopoverCmp_0_4.context, this.projectableNodes, null);
        var disposable_0 = this.renderer.listenGlobal('body', 'keyup', this.eventHandler(this._handle_keyup_0_0.bind(this)));
        this.init([].concat([this._el_0]), [this._el_0], [disposable_0], []);
        return this._appEl_0;
    };
    _View_PopoverCmp_Host0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import0.PopoverCmp) && (0 === requestNodeIndex))) {
            return this._PopoverCmp_0_4.context;
        }
        return notFoundResult;
    };
    _View_PopoverCmp_Host0.prototype.detectChangesInternal = function (throwOnChange) {
        this._PopoverCmp_0_4.detectChangesInternal(this, this._el_0, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_PopoverCmp_Host0.prototype._handle_keyup_0_0 = function ($event) {
        this._appEl_0.componentView.markPathToRootAsCheckOnce();
        var pd_0 = (this._PopoverCmp_0_4.context._keyUp($event) !== false);
        return (true && pd_0);
    };
    return _View_PopoverCmp_Host0;
}(import1.AppView));
function viewFactory_PopoverCmp_Host0(viewUtils, parentInjector, declarationEl) {
    if ((renderType_PopoverCmp_Host === null)) {
        (renderType_PopoverCmp_Host = viewUtils.createRenderComponentType('', 0, import13.ViewEncapsulation.None, [], {}));
    }
    return new _View_PopoverCmp_Host0(viewUtils, parentInjector, declarationEl);
}
export var PopoverCmpNgFactory = new import14.ComponentFactory('ion-popover', viewFactory_PopoverCmp_Host0, import0.PopoverCmp);
var styles_PopoverCmp = [];
var renderType_PopoverCmp = null;
var _View_PopoverCmp0 = (function (_super) {
    __extends(_View_PopoverCmp0, _super);
    function _View_PopoverCmp0(viewUtils, parentInjector, declarationEl) {
        _super.call(this, _View_PopoverCmp0, renderType_PopoverCmp, import6.ViewType.COMPONENT, viewUtils, parentInjector, declarationEl, import7.ChangeDetectorStatus.CheckAlways);
    }
    _View_PopoverCmp0.prototype.createInternal = function (rootSelector) {
        var parentRenderNode = this.renderer.createViewRoot(this.declarationAppElement.nativeElement);
        this._viewQuery_viewport_0 = new import15.QueryList();
        this._el_0 = this.renderer.createElement(parentRenderNode, 'ion-backdrop', null);
        this.renderer.setElementAttribute(this._el_0, 'disable-activated', '');
        this.renderer.setElementAttribute(this._el_0, 'role', 'presentation');
        this.renderer.setElementAttribute(this._el_0, 'tappable', '');
        this._Backdrop_0_3 = new import16.Wrapper_Backdrop(this.parentInjector.get(import17.GestureController), new import9.ElementRef(this._el_0), this.renderer);
        this._el_1 = this.renderer.createElement(parentRenderNode, 'div', null);
        this.renderer.setElementAttribute(this._el_1, 'class', 'popover-wrapper');
        this._el_2 = this.renderer.createElement(this._el_1, 'div', null);
        this.renderer.setElementAttribute(this._el_2, 'class', 'popover-arrow');
        this._el_3 = this.renderer.createElement(this._el_1, 'div', null);
        this.renderer.setElementAttribute(this._el_3, 'class', 'popover-content');
        this._el_4 = this.renderer.createElement(this._el_3, 'div', null);
        this.renderer.setElementAttribute(this._el_4, 'class', 'popover-viewport');
        this._el_5 = this.renderer.createElement(this._el_4, 'div', null);
        this.renderer.setElementAttribute(this._el_5, 'nav-viewport', '');
        this._appEl_5 = new import3.AppElement(5, 4, this, this._el_5);
        this._expr_1 = import7.UNINITIALIZED;
        var disposable_0 = this.renderer.listen(this._el_0, 'click', this.eventHandler(this._handle_click_0_0.bind(this)));
        this._viewQuery_viewport_0.reset([this._appEl_5.vcRef]);
        this.context._viewport = this._viewQuery_viewport_0.first;
        this.init([], [
            this._el_0,
            this._el_1,
            this._el_2,
            this._el_3,
            this._el_4,
            this._el_5
        ], [disposable_0], []);
        return null;
    };
    _View_PopoverCmp0.prototype.injectorGetInternal = function (token, requestNodeIndex, notFoundResult) {
        if (((token === import18.Backdrop) && (0 === requestNodeIndex))) {
            return this._Backdrop_0_3.context;
        }
        return notFoundResult;
    };
    _View_PopoverCmp0.prototype.detectChangesInternal = function (throwOnChange) {
        this._Backdrop_0_3.detectChangesInternal(this, this._el_0, throwOnChange);
        this.detectContentChildrenChanges(throwOnChange);
        var currVal_1 = !this.context.d.showBackdrop;
        if (import4.checkBinding(throwOnChange, this._expr_1, currVal_1)) {
            this.renderer.setElementClass(this._el_0, 'hide-backdrop', currVal_1);
            this._expr_1 = currVal_1;
        }
        this.detectViewChildrenChanges(throwOnChange);
    };
    _View_PopoverCmp0.prototype.destroyInternal = function () {
        this._Backdrop_0_3.context.ngOnDestroy();
    };
    _View_PopoverCmp0.prototype._handle_click_0_0 = function ($event) {
        this.markPathToRootAsCheckOnce();
        var pd_0 = (this.context._bdClick() !== false);
        return (true && pd_0);
    };
    return _View_PopoverCmp0;
}(import1.AppView));
export function viewFactory_PopoverCmp0(viewUtils, parentInjector, declarationEl) {
    if ((renderType_PopoverCmp === null)) {
        (renderType_PopoverCmp = viewUtils.createRenderComponentType('', 0, import13.ViewEncapsulation.None, styles_PopoverCmp, {}));
    }
    return new _View_PopoverCmp0(viewUtils, parentInjector, declarationEl);
}
